File: sealer2100/jub-sdk-android/app/src/main/java/com/legendwd/hyperpay/onchain/hardware/jubnative/protocol/ProtocolEncoder.java
package com.legendwd.hyperpay.onchain.hardware.jubnative.protocol;

import android.util.Log;

import com.legendwd.hyperpay.onchain.hardware.jubnative.utils.StrUtil;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

public class ProtocolEncoder {
    private static final String TAG = "ProtocolEncoder";

    private static final int TREAZOR_PACKET_SIZE = 64;
    private static final int ANDROID_BLE_PACKET_SIZE = 192;
    private static final int IOS_BLE_PACKET_SIZE = 128;

    // 协议头相关常量
    private static final byte TREAZOR_HEADER_FIRST = '#';
    private static final byte TREAZOR_HEADER_SECOND = '#';
    private static final byte TREAZOR_PACKET_PREFIX = '?';

    ///////////////////////////////////// trezor 协议编码 protobuf ///////////////////////////////////

    /**
     * trezor 协议编码 + 分包
     *
     * @param messageType
     * @param data
     * @return
     */
    public static byte[] encodeProtocol(int messageType, byte[] data) {
        if (data == null) {
            logError("encodeProtocol: 输入数据不能为 null");
            return new byte[0];
        }

        byte[] pbBytes = encode(messageType, data);
//        logDebug("encodeProtocol - pbBytes: " + StrUtil.byteArr2HexStr(pbBytes));

        List<byte[]> packets = splitDataWithHeader(pbBytes);
        byte[] cmd = convertListToByteArray(packets);
        logDebug("encodeProtocol - cmd(padding): " + StrUtil.byteArr2HexStr(cmd));

        return cmd;
    }

    /**
     * trezor 协议编码
     *
     * 首包格式：?##<msg type><data len><data>
     * 其他包格式：<?><data><补零>
     *
     * @param data
     * @return
     */
    static byte[] encode(int messageType, byte[] data) {
        if (data == null) {
            logError("encode: 输入数据不能为 null");
            return new byte[0];
        }

        int bufferSize = calculateEncodeBufferSize(data);
        ByteBuffer buffer = ByteBuffer.allocate(bufferSize);

        buffer.put(TREAZOR_HEADER_FIRST)
                .put(TREAZOR_HEADER_SECOND)
                .putShort((short) messageType)
                .putInt(data.length)
                .put(data);

        return buffer.array();
    }

    /**
     * 计算编码所需的 ByteBuffer 大小
     *
     * @param data 待编码的数据
     * @return 所需的 ByteBuffer 大小
     */
    private static int calculateEncodeBufferSize(byte[] data) {
        // 2 个字节的协议头 + 2 个字节的消息类型 + 4 个字节的数据长度 + 数据本身长度
        return 2 + 2 + 4 + data.length;
    }


    /**
     * 将数据按每包 64 字节分割，每包首字节为 '?'，不足的补零
     *
     * @param data 待分割的原始数据
     * @return 分割后的数据包列表
     */
    static List<byte[]> splitDataWithHeader(byte[] data) {
        if (data == null) {
            logError("splitDataWithHeader: 输入数据不能为 null");
            return new ArrayList<>();
        }

        List<byte[]> packetList = new ArrayList<>();
        int offset = 0;

        while (offset < data.length) {
            byte[] packet = createPacket(data, offset);
            packetList.add(packet);
            offset += TREAZOR_PACKET_SIZE - 1;
        }

        return packetList;
    }

    /**
     * 创建单个分包
     *
     * @param data   原始数据
     * @param offset 当前偏移量
     * @return 单个分包
     */
    private static byte[] createPacket(byte[] data, int offset) {
        // 每包 64 字节，首字节为 '?'
        byte[] packet = new byte[TREAZOR_PACKET_SIZE];
        packet[0] = TREAZOR_PACKET_PREFIX;

        int copyLength = Math.min(data.length - offset, TREAZOR_PACKET_SIZE - 1);
        System.arraycopy(data, offset, packet, 1, copyLength);

        return packet;
    }

    /**
     * 使用 ByteBuffer 将 List<byte[]> 转换为一个 byte[] 数组
     *
     * @param list 包含多个 byte[] 的列表
     * @return 合并后的 byte[] 数组，如果输入列表为 null 或空，返回空数组
     */
    static byte[] convertListToByteArray(List<byte[]> list) {
        if (list == null || list.isEmpty()) {
            return new byte[0];
        }

        int totalLength = 0;
        for (byte[] bytes : list) {
            if (bytes != null) {
                totalLength += bytes.length;
            }
        }

        ByteBuffer buffer = ByteBuffer.allocate(totalLength);

        for (byte[] bytes : list) {
            if (bytes != null) {
                buffer.put(bytes);
            }
        }

        return buffer.array();
    }

    ////////////////////////////////////////// 蓝牙分包 /////////////////////////////////////////////

    /**
     * 分割数据为指定大小的包
     *
     * @param data
     * @param sliceSize android: 192 bytes, ios: 128 bytes
     * @return
     */
    static List<byte[]> slice(byte[] data, int sliceSize) {
        if (data == null || sliceSize <= 0) {
            logError("slice: 输入数据不能为 null 且分片大小必须大于 0");
            return new ArrayList<>();
        }

        // 分片数
        int pieces = (int) Math.floor((data.length - 1) / sliceSize) + 1;
        List<byte[]> sliceLists = new ArrayList<>(pieces);
        int offset = 0;

        while (offset < data.length) {
            int packetLength = Math.min(data.length - offset, sliceSize);
            byte[] buffer = new byte[packetLength];
            System.arraycopy(data, offset, buffer, 0, packetLength);
            sliceLists.add(buffer);
            offset += packetLength;
        }
        return sliceLists;
    }

    /**
     * 分割数据为 192 bytes 大小的包
     *
     * @param data
     * @return
     */
    public static List<byte[]> slice(byte[] data) {
        return slice(data, ANDROID_BLE_PACKET_SIZE);
    }

    private static void logError(String message) {
        Log.e(TAG, message);
    }

    private static void logDebug(String message) {
        Log.d(TAG, message);
    }

}

File: sealer2100/jub-sdk-cxx/deps/trezorProtocol/ProtocolEncoder.cpp
#include "ProtocolEncoder.h"
#include <iostream>
#include <algorithm>
#include <cstring>

#ifdef __ANDROID__
#include "utils.h"
#include "utils/logUtils.h"
#endif

void ProtocolEncoder::logError(const std::string& message) {
    std::cerr << "ProtocolEncoder: " << message << std::endl;
}

void ProtocolEncoder::logDebug(const std::string& message) {
    std::cout << "ProtocolEncoder: " << message << std::endl;
}

int ProtocolEncoder::calculateEncodeBufferSize(const std::vector<uint8_t>& data) {
    // 2 个字节的协议头 + 2 个字节的消息类型 + 4 个字节的数据长度 + 数据本身长度
    return 2 + 2 + 4 + data.size();
}

std::vector<uint8_t> ProtocolEncoder::encode(int messageType, const std::vector<uint8_t>& data) {
    int bufferSize = calculateEncodeBufferSize(data);
    std::vector<uint8_t> buffer(bufferSize);

    buffer[0] = TREAZOR_HEADER_FIRST;
    buffer[1] = TREAZOR_HEADER_SECOND;
    buffer[2] = static_cast<uint8_t>(messageType >> 8);
    buffer[3] = static_cast<uint8_t>(messageType & 0xFF);
    buffer[4] = static_cast<uint8_t>(data.size() >> 24);
    buffer[5] = static_cast<uint8_t>((data.size() >> 16) & 0xFF);
    buffer[6] = static_cast<uint8_t>((data.size() >> 8) & 0xFF);
    buffer[7] = static_cast<uint8_t>(data.size() & 0xFF);
    std::memcpy(buffer.data() + 8, data.data(), data.size());

    return buffer;
}
std::vector<uint8_t> ProtocolEncoder::encodeProtocol(int messageType, const std::string &data){
    std::vector<uint8_t> dataBytes(data.begin(), data.end());
    return ProtocolEncoder::encodeProtocol(messageType, dataBytes);
}

std::vector<uint8_t> ProtocolEncoder::encodeProtocol(int messageType, const std::vector<uint8_t>& data) {
#ifdef __ANDROID__
#ifdef DEBUG
    std::string hexBuffer = toHexString(data);
    LOG_INF("[cmd] 待编码指令 : messageType: %d, data: %s", messageType, hexBuffer.c_str());
#endif
#endif

    std::vector<uint8_t> pbBytes = encode(messageType, data);

    std::vector<std::vector<uint8_t>> packets = splitDataWithHeader(pbBytes);
    std::vector<uint8_t> cmd = convertListToByteArray(packets);
    // logDebug("encodeProtocol - cmd(padding): " + HexString.byteArrayToHex(cmd));

    return cmd;
}

std::vector<std::vector<uint8_t>> ProtocolEncoder::splitDataWithHeader(const std::vector<uint8_t>& data) {
    if (data.empty()) {
        logError("splitDataWithHeader: 输入数据不能为 null");
        return {};
    }

    std::vector<std::vector<uint8_t>> packetList;
    int offset = 0;

    while (offset < data.size()) {
        std::vector<uint8_t> packet = createPacket(data, offset);
        packetList.push_back(packet);
        offset += TREAZOR_PACKET_SIZE - 1;
    }

    return packetList;
}

std::vector<uint8_t> ProtocolEncoder::createPacket(const std::vector<uint8_t>& data, int offset) {
    // 每包 64 字节，首字节为 '?'
    std::vector<uint8_t> packet(TREAZOR_PACKET_SIZE, 0);
    packet[0] = TREAZOR_PACKET_PREFIX;

    int copyLength = std::min(static_cast<int>(data.size() - offset), TREAZOR_PACKET_SIZE - 1);
    std::memcpy(packet.data() + 1, data.data() + offset, copyLength);

    return packet;
}

std::vector<uint8_t> ProtocolEncoder::convertListToByteArray(const std::vector<std::vector<uint8_t>>& list) {
    if (list.empty()) {
        return {};
    }

    size_t totalLength = 0;
    for (const auto& bytes : list) {
        totalLength += bytes.size();
    }

    std::vector<uint8_t> buffer(totalLength);
    size_t offset = 0;
    for (const auto& bytes : list) {
        std::memcpy(buffer.data() + offset, bytes.data(), bytes.size());
        offset += bytes.size();
    }

    return buffer;
}

std::vector<std::vector<uint8_t>> ProtocolEncoder::slice(const std::vector<uint8_t>& data, int sliceSize) {
    if (data.empty() || sliceSize <= 0) {
        logError("slice: 输入数据不能为 null 且分片大小必须大于 0");
        return {};
    }

    std::vector<std::vector<uint8_t>> sliceLists;
    int offset = 0;

    while (offset < data.size()) {
        int packetLength = std::min(static_cast<int>(data.size() - offset), sliceSize);
        std::vector<uint8_t> buffer(data.begin() + offset, data.begin() + offset + packetLength);
        sliceLists.push_back(buffer);
        offset += packetLength;
    }
    return sliceLists;
}

std::vector<std::vector<uint8_t>> ProtocolEncoder::slice(const std::vector<uint8_t>& data) {
    return slice(data, ANDROID_BLE_PACKET_SIZE);
}
